src/test/java/picocli/CommandModelTransformersTest.java

```diff
@@ -0,0 +1,203 @@
+    }
+
+    @Test
+    public void testNestedSubcommandFiltering() {
+        System.setProperty("disable_aa", "true");
+
+        CommandLine cmd = new CommandLine(new Main());
+
+        // there is only function annotated subcommand now
+        assertArrayEquals(new String[]{"ab"}, cmd.getSubcommands().get("a").getSubcommands().keySet().toArray(new String[0]));
+    }
+
+    @Test
+    public void testFunctionAnnotatedCommandFiltering() {
+        System.setProperty("disable_ab", "true");
+
+        CommandLine cmd = new CommandLine(new Main());
+
+        // there is only class annotated subcommand now
+        assertArrayEquals(new String[]{"aa"}, cmd.getSubcommands().get("a").getSubcommands().keySet().toArray(new String[0]));
+    }
+
+    @Test(expected = CommandLine.UnmatchedArgumentException.class)
+    public void programmaticAPICommandFiltering() {
+
+        @CommandLine.Command(name="a")
+        class A implements Callable<Integer> {
+            public Integer call() throws Exception {
+                return 0;
+            }
+        }
+        @CommandLine.Command(name="b")
+        class B implements Callable<Integer> {
+            public Integer call() throws Exception {
+                return 0;
+            }
+        }
+        @CommandLine.Command(name="c")
+        class C implements Callable<Integer> {
+            public Integer call() throws Exception {
+                return 0;
+            }
+        }
+        @CommandLine.Command(name="d")
+        class D implements Callable<Integer> {
+            public Integer call() throws Exception {
+                return 2;
+            }
+        }
+
+
+        class MT implements CommandLine.IModelTransformer {
+            public CommandLine.Model.CommandSpec transform(CommandLine.Model.CommandSpec commandSpec) {
+                commandSpec.removeSubcommand("a");
+                commandSpec.removeSubcommand("c");
+
+                commandSpec.addSubcommand("d", CommandLine.Model.CommandSpec.forAnnotatedObject(new D()));
+
+                return commandSpec;
+            }
+        }
+
+        CommandLine.IModelTransformer mt = new MT();
+
+        CommandLine.Model.CommandSpec commandSpec = CommandLine.Model.CommandSpec
+                .create()
+                .modelTransformer(mt)
+                .addSubcommand("a", new CommandLine(new A()))
+                .addSubcommand( "b", new CommandLine(new B()))
+                .addSubcommand("c", new CommandLine(new C()));
+
+        assertEquals(mt, commandSpec.modelTransformer());
+
+        final CommandLine cl = new CommandLine(commandSpec);
+
+        assertArrayEquals(new String[]{"b", "d"}, cl.getSubcommands().keySet().toArray(new String[0]));
+
+        assertEquals(2, cl.execute("d"));
+
+        assertEquals(0, cl.execute("b"));
+
+        // should throw unmatched argument exception
+        cl.parseArgs("b");
+
+        // should throw unmatched argument exception
+        cl.parseArgs("c");
```

Comment:
Is this statement reached, given that `parseArgs("b")` will throw an exception?

Reply:
Nope, I'll fix it.